apps/storefront/src/pages/Login/useLogout.tsx

```diff
@@ -0,0 +1,57 @@
+import { useCallback } from 'react';
+import { B2BEvent } from '@b3/hooks';
+
+import { endUserMasqueradingCompany, superAdminEndMasquerade } from '@/shared/service/b2b';
+import { bcLogoutLogin } from '@/shared/service/bc';
+import { clearMasqueradeCompany, useAppDispatch, useAppSelector } from '@/store';
+import b2bLogger from '@/utils/b3Logger';
+import { logoutSession } from '@/utils/b3logout';
+
+export const useMasquerade = () => {
+  const isAgenting = useAppSelector(({ b2bFeatures }) => b2bFeatures.masqueradeCompany.isAgenting);
+  const salesRepCompanyId = useAppSelector(({ b2bFeatures }) => b2bFeatures.masqueradeCompany.id);
+  const storeDispatch = useAppDispatch();
+
+  return useCallback(async () => {
+    if (isAgenting) {
+      await superAdminEndMasquerade(Number(salesRepCompanyId));
+      storeDispatch(clearMasqueradeCompany());
+    }
+  }, [salesRepCompanyId, storeDispatch, isAgenting]);
+};
+
+export const useEndCompanyMasquerading = () => {
+  const { selectCompanyHierarchyId } = useAppSelector(
+    ({ company }) => company.companyHierarchyInfo,
+  );
+
+  return useCallback(async () => {
+    if (selectCompanyHierarchyId) {
+      await endUserMasqueradingCompany();
+    }
+  }, [selectCompanyHierarchyId]);
+};
+
+export const useLogout = () => {
+  const endMasquerade = useMasquerade();
+  const endCompanyMasquerading = useEndCompanyMasquerading();
+
+  return async () => {
+    try {
+      const { result } = (await bcLogoutLogin()).data.logout;
+
+      if (result !== 'success') {
```

Comment:
can the promise resolve to any other state?

if not, I'd consider chaining with a `then` 🍹

Reply:
this is not the resolve state of the promise, but rather the response from the server. By looking at the code I'm assuming there's a scenario where it can fail.